lite/provider.go

```diff
@@ -30,3 +33,105 @@ type PersistentProvider interface {
+	// will return this height.
+	//
+	// NOTE: Behavior with concurrent requests is undefined.  To make
+	// concurrent calls safe, look at the struct `ConcurrentUpdatingProvider`.
+	UpdateToHeight(chainID string, height int64) error
+}
+
+//----------------------------------------
+
+type concurrentProvider struct {
+	UpdatingProvider
+
+	// pending map to synchronize concurrent verification requests
+	mtx                  sync.Mutex
+	pendingVerifications map[pendingKey]*pendingResult
+}
+
+// convenience to create the key for the lookup map
+type pendingKey struct {
+	chainID string
+	height  int64
+}
+
+// used to cache the result from underlying UpdatingProvider.
+type pendingResult struct {
+	wait chan struct{}
+	err  error // cached result.
+}
+
+func NewConcurrentUpdatingProvider(up UpdatingProvider) concurrentProvider {
+	return concurrentProvider{
+		UpdatingProvider:     up,
+		pendingVerifications: make(map[pendingKey]*pendingResult),
+	}
+}
+
+// Returns the unique pending request for all identical calls to
+// joinConcurrency(chainID,height), and returns true for isFirstCall only for the
+// first call, which should call the returned callback w/ results if any.
+// The callback must be called, otherwise there will be memory leaks.
+// Other subsequent calls should just return uniq.err.
+// NOTE: This is a separate function, primarily to make mtx unlocking more obviously safe via defer.
+func (cp concurrentProvider) joinConcurrency(chainID string, height int64) (uniq *pendingResult, isFirstCall bool, callback func(error)) {
+	cp.mtx.Lock()
+	defer cp.mtx.Unlock()
+
+	var pkey = pendingKey{chainID, height}
+	cp.mtx.Lock()
+	if uniq = cp.pendingVerifications[pkey]; uniq != nil {
+		cp.mtx.Unlock()
+		<-uniq.wait // uniq.wait is of type `chan struct{}`
+		return uniq, false, nil
+	} else {
+		uniq = &pendingResult{wait: make(chan struct{}), err: nil}
+		cp.pendingVerifications[pkey] = uniq
+		cp.mtx.Unlock()
+		// The caller must call this, otherwise there will be memory leaks.
+		return uniq, true, func(err error) {
+			// NOTE: other result parameters can be added here.
+			uniq.err = err
+			// *After* setting the results, *then* call close(uniq.wait).
+			close(uniq.wait)
+			cp.mtx.Lock() // temporarily acquire lock to remove this iteem
+			delete(cp.pendingVerifications, pkey)
+			cp.mtx.Unlock() // and release lock
+		}
+	}
+}
+
+func (cp concurrentProvider) UpdateToHeight(chainID string, height int64) error {
+
+	// Performs synchronization for multi-threads verification at the same height.
+	var presult *pendingResult
+	var isFirstCall bool
+	var callback func(error)
+	presult, isFirstCall, callback = cp.joinConcurrency(chainID, height)
+
+	if isFirstCall {
+		var err error
+		// Use a defer in case UpdateToHeight itself fails.
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("Recovered from panic: %v", r)
+			}
+			callback(err)
```

Comment:
what if there's no error? we don't call the `callback`

comment in `joinConcurrency` states "// The caller must call this, otherwise there will be memory leaks."

Reply:
The `defer` will be executed on `isFirstCall`. Independent from if there is an error or not ...